orchagent/cbf/cbfnhgorch.cpp

```diff
@@ -0,0 +1,819 @@
+            throw std::logic_error("Syncing already synced NHG member");
+        }
+
+        /*
+         * Check if the group exists in NhgOrch.
+         */
+        if (!gNhgOrch->hasNhg(key))
+        {
+            SWSS_LOG_ERROR("Next hop group %s in CBF next hop group %s does "
+                            "not exist", key.c_str(), m_key.c_str());
+            return false;
+        }
+
+        const auto &nhg = gNhgOrch->getNhg(key);
+
+        /*
+         * Check if the group is synced.
+         */
+        if (!nhg.isSynced())
+        {
+            SWSS_LOG_ERROR("Next hop group %s in CBF next hop group %s is not"
+                            " synced",
+                            key.c_str(), m_key.c_str());
+            return false;
+        }
+
+        /* Create the SAI attributes for syncing the NHG as a member. */
+        auto attrs = createNhgmAttrs(nhgm);
+
+        bulker.create_entry(&nhgm_ids[key],
+                            (uint32_t)attrs.size(),
+                            attrs.data());
+    }
+
+    /* Flush the bulker to perform the sync. */
+    bulker.flush();
+
+    /* Iterate over the synced members and set their SAI ID. */
+    bool success = true;
+
+    for (const auto &member : nhgm_ids)
+    {
+        if (member.second == SAI_NULL_OBJECT_ID)
+        {
+            SWSS_LOG_ERROR("Failed to create CBF next hop group %s member %s",
+                            m_key.c_str(), member.first.c_str());
+            success = false;
+        }
+        else
+        {
+            m_members.at(member.first).sync(member.second);
+
+            const auto &nhg = gNhgOrch->getNhg(member.first);
+            /*
+             * If the member is temporary, store it in order to check if it is
+             * promoted at some point.
+             */
+            if (nhg.isTemp())
+            {
+                m_temp_nhgs.emplace(member.first, nhg.getId());
+            }
+        }
+    }
+
+    return success;
+}
+
+/*
+ * Purpose: Create a vector with the SAI attributes for syncing a next hop
+ *          group member over SAI.  The caller is reponsible of filling in the
+ *          index attribute.
+ *
+ * Params:  IN nhgm - The next hop group member to sync.
+ *
+ * Returns: The vector containing the SAI attributes.
+ */
+vector<sai_attribute_t> CbfNhg::createNhgmAttrs(const CbfNhgMember &nhgm) const
+{
+    SWSS_LOG_ENTER();
+
+    if (!isSynced() || (nhgm.getNhgId() == SAI_NULL_OBJECT_ID))
+    {
+        SWSS_LOG_ERROR("CBF next hop group %s or next hop group %s are not synced",
+                       to_string().c_str(), nhgm.to_string().c_str());
+        throw logic_error("CBF next hop group member attributes data is insufficient");
+    }
+
+    vector<sai_attribute_t> attrs;
+    sai_attribute_t attr;
+
+    /* Fill in the group ID. */
+    attr.id = SAI_NEXT_HOP_GROUP_MEMBER_ATTR_NEXT_HOP_GROUP_ID;
+    attr.value.oid = m_id;
+    attrs.push_back(attr);
+
+    /* Fill in the next hop ID. */
```

Comment:
Do we check if this next_hop is reachable or not similar to RouteOrch::addNextHopGroup? Not sure if getNhgId method checks neighOrch for the valid nexthops.

Reply:
The next hop is always a next hop group owned by nhgOrch, and we have  a check higher in the function to check if it's been programmed over the SAI (at which point it must be usable)